FVM Open RFP (#592)
filecoin-project/devgrants: rfps/rfp-fvm.md

# FVM Tooling & Infrastructure RFP

## Overview

We are seeking proposals for developer and infrastructure tooling to extend the functionality and usability of the forthcoming [Filecoin Virtual Machine](https://fvm.filecoin.io) (FVM).

The Filecoin Virtual Machine (FVM) combines smart contracts with co-location of storage and compute capabilities.

The FVM is a WASM-based polyglot execution environment for IPLD data. It is designed to support native Filecoin actors written in languages that compile to WASM, as well as smart contracts written for foreign runtimes including the Ethereum Virtual Machine (EVM), Secure EcmaScript (SES), and eBPF.

You can also consider contributing via the [FVM Bug Bounties](https://immunefi.com/bounty/filecoin/) program.

## Description

We seek to fund the creation of world-class tools that enable future developers to build on the FVM easily and seamlessly.

We are therefore seeking proposals for various developer and infrastructure tooling to extend the functionality and usability of the FVM.

> These projects are essential to build the minimum set of tools that we would need to provide in order to have an end-to-end builders process for people who are deploying smart contracts on FVM.

Reviewer: Rephrase suggestion: 

These projects are essential for providing the minimum set of tools needed in order to have an end-to-end process for builders deploying smart contracts on FVM.

The FVM is a WASM-based polyglot execution environment for IPLD data. It is designed to support native Filecoin actors written in languages that compile to WASM, as well as smart contracts written for foreign runtimes including the Ethereum Virtual Machine (EVM), Secure EcmaScript (SES), and eBPF.

Refer to the resources section for more information on the FVM reference and specification, details on the possibilities FVM will unlock and the current roadmap.

You can also consider contributing via the open [FVM Bug Bounties page here](https://immunefi.com/bounty/filecoin/)

## Description

The FVM is an integral and high-priority project for extending the functionality of Filecoin by bringing storage solutions to the blockchain and executing computation jobs or data pipelines on globally distributed data by pushing compute to the edges and coordinating its execution.

Examples of future ideas include: async programming via continuations/futures, capability-orientation, time-locked transactions (using external beacons), parallel execution, native primitives for cross-chain interoperability, computation correctness proofs, and more.

> We want to build the best tech and tooling available to help enable future developers to build the use cases they envision on the FVM easily and seamlessly.

Reviewer 2: ```suggestion
We seek to fund the creation of world-class tools that enable future developers to build on the FVM easily and seamlessly.
```

We are therefore seeking proposals for various developer and infrastructure tooling to extend the functionality and usability of the FVM.

These projects are essential to build the minimum set of tools that we would need to provide in order to have an end-to-end builders process for people who are deploying smart contracts on FVM.

## Project Types

The projects needed include (but are not limited to):

- SDK's
  - High-level Rust (abstraction layer on the current FVM Rust reference code)
  - Go
  - AssemblyScript
  - Kotlin
  - Swift
  - Others

- Native Tooling & Developer Tooling (see resources for a more extensive list)
  - Local runtime tools and developer environments (like ethereum's Hardhat)
  - Debugging & Testing tools
  - Deployment flow tooling
  - Testnets, Faucets, RPC Endpoints & local nodes
  - Smart contract playgrounds and standards (for example OpenZeppelin and Remix type applications)
  - Foreign Runtime Bridges (including EVM) & Oracles
  - Wallets and wallet integrations

## Deliverables

> Teams should provide a clear outline of the scope of the work and milestones along with a time-frame for completing these with a monetary funding amount requested for each. These can be in the form of a research project and associated reporting, a prototype project or MVP map and associated outcomes and specifications.

Reviewer 2: ```suggestion
Teams should provide a clear outline of the scope of the work, milestones, timeframe, and funding requested. These can be in the form of a research project and associated reporting, or a prototype project or MVP map and associated outcomes and specifications.
```

### Proposal Guidelines

We evaluate proposals based on:

- The quality of the proposal and near-term value to the Filecoin ecosystem
- Your team’s technical experience, open source contributions and interest in diving into how Filecoin works
- Likelihood that your team will continue to support the Filecoin ecosystem long-term and commitment to maintaining your project for one year

Note that all proposals must be open sourced via MIT and Apache2 licenses.

## Resources

The following resources may be helpful:

#### General FVM Resources

- [FVM Website with Roadmap](https://fvm.filecoin.io)
- [FVM Reference Implementation](https://github.com/filecoin-project/ref-fvm)
- [FVM Specification](https://github.com/filecoin-project/fvm-specs)
- [FIP Proposal](https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0031.md#coordinated-testnets)
- [FVM Early Builders Program Information & Meeting Recordings](https://pl-strflt.notion.site/Welcome-to-the-FVM-Foundry-Early-Builders-Program-0c21aa082a3a48b5b9262d122204d141)
- [Filecoin Project FVM Slack Channel](https://filecoinproject.slack.com/archives/C029MT4PQB1)

#### Application Examples

> - Example proposoal from Polyphene: [High-level Rust SDK](https://github.com/filecoin-project/devgrants/issues/562)

Reviewer 2: ```suggestion
- Example proposal from Polyphene: [High-level Rust SDK](https://github.com/filecoin-project/devgrants/issues/562)
```

## Overview

We are seeking proposals for various developer and infrastructure tooling to extend the functionality and usability of the [Filecoin Virtual Machine](https://fvm.filecoin.io) (FVM).

The Filecoin Virtual Machine (FVM) aims to combine smart contracts with co-location of storage and compute capabilities.

The FVM is a WASM-based polyglot execution environment for IPLD data. It is designed to support native Filecoin actors written in languages that compile to WASM, as well as smart contracts written for foreign runtimes including the Ethereum Virtual Machine (EVM), Secure EcmaScript (SES), and eBPF.

Refer to the resources section for more information on the FVM reference and specification, details on the possibilities FVM will unlock and the current roadmap.

You can also consider contributing via the open [FVM Bug Bounties page here](https://immunefi.com/bounty/filecoin/)

## Description

The FVM is an integral and high-priority project for extending the functionality of Filecoin by bringing storage solutions to the blockchain and executing computation jobs or data pipelines on globally distributed data by pushing compute to the edges and coordinating its execution.

> Examples of future ideas include: async programming via continuations/futures, capability-orientation, time-locked transactions (using external beacons), parallel execution, native primitives for cross-chain interoperability, computation correctness proofs, and more.

Reviewer 3: @DeveloperAlly - This paragraph is about the longer-term future of the FVM and is a bit confusing in the context of this RFP description for infra and dev tooling. I recommend removing it.

Author: I thought it might be good for context. As most of the grants will probably come through us anyway. I'm also fine to just remove it

Author: Removed

## Description

The FVM is an integral and high-priority project for extending the functionality of Filecoin by bringing storage solutions to the blockchain and executing computation jobs or data pipelines on globally distributed data by pushing compute to the edges and coordinating its execution.

Examples of future ideas include: async programming via continuations/futures, capability-orientation, time-locked transactions (using external beacons), parallel execution, native primitives for cross-chain interoperability, computation correctness proofs, and more.

We want to build the best tech and tooling available to help enable future developers to build the use cases they envision on the FVM easily and seamlessly.

We are therefore seeking proposals for various developer and infrastructure tooling to extend the functionality and usability of the FVM.

These projects are essential to build the minimum set of tools that we would need to provide in order to have an end-to-end builders process for people who are deploying smart contracts on FVM.

## Project Types

The projects needed include (but are not limited to):

- SDKs

>   - High-level Rust

Reviewer 2: Can you add more detail around what high-level Rust means? (Unless it's something the target audience would immediately understand.)

Author: Yea - we already have a proposal for that one in, so it's kind of already a specific case

Author: Also I removed the comma because it kept messing up the formatting! shrug

Author: Fixed (both)!
